### PR TITLE
Feature/pareto

### PR DESCRIPTION
**Pull request by christophernorth**
Added new Pareto Vertex for the Pareto Distribution.
Also fixed a bug (Issue #183) in the Exponential Vertex where Scalar params caused us to ignore all but the first x value in our input tensor for dLogProb (and added a missing test for this Distribution at the same time).
Also made a small change to the Cauchy tests to cover a wider area (to ensure non-zero probs for some of the range).

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Exponential.java [hunk 1]
```diff
@@ -34,7 +34,7 @@ public DoubleTensor logProb(DoubleTensor x) {
 
     @Override
     public Diffs dLogProb(DoubleTensor x) {
-        final DoubleTensor dLogPdx = lambda.reciprocal().unaryMinusInPlace();
+        final DoubleTensor dLogPdx = DoubleTensor.zeros(x.getShape()).plusInPlace(lambda).reciprocalInPlace().unaryMinusInPlace();
```
Comment (icantsayrural) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Exponential.java]: Can we just do `minusInPlace` so we avoid calling `unaryMinusInPlace`?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/CauchyVertexTest.java [hunk 2]
```diff
@@ -109,7 +110,7 @@ public void dLogProbMatchesFiniteDifferenceCalculationFordPdlocation() {
 
         moveAlongDistributionAndTestGradientOnARangeOfHyperParameterValues(
             Nd4jDoubleTensor.scalar(1.0),
-            Nd4jDoubleTensor.scalar(1.5),
```
Comment (icantsayrural) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/CauchyVertexTest.java]: Why the change?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Pareto.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Override
+    public DoubleTensor logProb(DoubleTensor x) {
+        DoubleTensor result;
+
+        /*
+         * If we've been passed invalid values, then return Negative Infinity for all values, else just return the
+         * calculated value
+         */
+        if (!xm.greaterThan(0).allTrue() || !alpha.greaterThan(0).allTrue()) {
```
Comment (icantsayrural) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Pareto.java]: nit: `greaterThan(0.0)`

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Exponential.java (hunk 1 above)
Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Exponential.java]: Thanks!

Yep we can - I'd just changed the zeros stuff so we got the Tensor created, but looking at the equation, yeah just doing a minus is better - good catch.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Pareto.java (hunk 3 above)
Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Pareto.java]: Fixed

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/CauchyVertexTest.java (hunk 2 above)
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/CauchyVertexTest.java]: This was just to match the other test and run over the full range of the uniform.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/ParetoVertex.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
+
+public class ParetoVertex extends DoubleVertex implements ProbabilisticDouble {
+    private final DoubleVertex alpha;
+    private final DoubleVertex xm;
+
+    /**
+     * Provides a Vertex implementing the Pareto Distribution.
+     *
+     * Requires at least on of location (xm) or scale(alpha) that match a proposed tensor shape
```
Comment (GeorgeNash) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/ParetoVertex.java]: `Requires at least one`

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ExponentialVertexTest.java [hunk 5]
```diff
@@ -73,6 +75,18 @@ public void matchesKnownDerivativeLogDensityOfScalar() {
         assertEquals(exponentialLogDiff.dPdx, actual.withRespectTo(tensorExponentialVertex.getId()).scalar(), 1e-5);
     }
 
+    @Test
+    public void matchesKnownDerivativeLogDensityOfVector() {
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ExponentialVertexTest.java]: Nice work

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ParetoVertexTest.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.tensor.dbl.Nd4jDoubleTensor;
+import io.improbable.keanu.vertices.ConstantVertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
+
+public class ParetoVertexTest {
+    private KeanuRandom random;
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ParetoVertexTest.java]: nit: white space

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ParetoVertexTest.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            random);
+    }
+
+    @Test
+    public void inferHyperParamsFromSamples() {
+        double trueXm = 5.0;
+
+        /*
+         * Note, this value is set low as the Gradient Optimizer seems to struggle with values greater than this - we
+         * should revisit to see if the Optimizer can be Optimzed to deal better with this case.
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/ParetoVertexTest.java]: `we should revisit this with a more configurable Optimiser or a different Optimiser entirely!`
